master/internal/rm/kubernetesrm/job.go

```diff
@@ -0,0 +1,751 @@
+			required := strconv.Itoa(j.slotsPerPod)
+			var resourceName string
+			switch j.slotType {
+			case device.CPU:
+				resourceName = "CPU slots"
+			default:
+				resourceName = "GPUs"
+			}
+
+			msgText = fmt.Sprintf("Waiting for resources. %s %s are available, %s %s required",
+				available, resourceName, required, resourceName)
+		}
+	}
+
+	return msgText
+}
+
+func (j *job) getPodState(pod k8sV1.Pod) (cproto.State, error) {
+	switch pod.Status.Phase {
+	case k8sV1.PodPending:
+		// When pods are deleted, Kubernetes sometimes transitions pod statuses to pending
+		// prior to deleting them. In these cases we have observed that we do not always
+		// receive a PodFailed or a PodSucceeded message. We check if pods have a set pod
+		// deletion timestamp to see if this is the case.
+		if pod.ObjectMeta.DeletionTimestamp != nil {
+			j.syslog.Warn("marking pod as terminated due to deletion timestamp")
+			return cproto.Terminated, nil
+		}
+
+		for _, condition := range pod.Status.Conditions {
+			if condition.Type == k8sV1.PodScheduled && condition.Status == k8sV1.ConditionTrue {
+				return cproto.Starting, nil
+			}
+		}
+		return cproto.Assigned, nil
+
+	case k8sV1.PodRunning:
+		// Pods are in a running state as long as at least one container has not terminated.
+		// We check the status of the Determined containers directly to determine if they
+		// are still running.
+		containerStatuses, err := getDeterminedContainersStatus(
+			pod.Status.ContainerStatuses, j.containerNames)
+		if err != nil {
+			return "", err
+		}
+
+		for _, containerStatus := range containerStatuses {
+			if containerStatus.State.Terminated != nil {
+				return cproto.Terminated, nil
+			}
+		}
+
+		for _, containerStatus := range containerStatuses {
+			// Check that all Determined containers are running.
+			if containerStatus.State.Running == nil {
+				return cproto.Starting, nil
+			}
+		}
+
+		return cproto.Running, nil
+
+	case k8sV1.PodFailed, k8sV1.PodSucceeded:
+		return cproto.Terminated, nil
+
+	default:
+		return "", errors.Errorf(
+			"unexpected pod status %s for pod %s", pod.Status.Phase, pod.Name)
+	}
+}
+
+func getExitCodeAndMessage(pod *k8sV1.Pod, containerNames set.Set[string]) (*exitReason, error) {
+	if len(pod.Status.InitContainerStatuses) == 0 {
+		return nil, errors.Errorf(
+			"unexpected number of init containers when processing exit code for pod %s", pod.Name)
+	}
+
+	for _, initContainerStatus := range pod.Status.InitContainerStatuses {
+		if initContainerStatus.State.Terminated == nil {
+			continue
+		}
+		exitCode := initContainerStatus.State.Terminated.ExitCode
+		if exitCode != aproto.SuccessExitCode {
+			errMessage := fmt.Sprintf(
+				"container %s: %s", initContainerStatus.Name,
+				initContainerStatus.State.Terminated.Message,
+			)
+			return &exitReason{
+				code: int(exitCode),
+				msg:  errMessage,
+			}, nil
+		}
+	}
+
+	if len(pod.Status.ContainerStatuses) < len(containerNames) {
+		return nil, errors.Errorf(
```

Comment:
nit: why are you using errors.Errorf & fmt.Errorf interchangeably in this file? better to stick to one (fmt.Errorf) over the other?

Reply:
just converted the file to fmt.Errorf. it was only this way because i hadn't bother to refactor yet.